### ISSUE TITLE
Bug: Generated types are setting matches as never after a conditional rendering on React

### ISSUE THREAD

**Issue by BrunoQuaresma** [statelyai/xstate#3029]
### Description

Generated types are setting `matches` as `never` after a conditional rendering on React.

### Expected result

Do not mark `matches` as `never`.

### Actual result

Marking `matches` as `never`

<img width="889" alt="Screen Shot 2022-02-08 at 17 14 45" src="https://user-images.githubusercontent.com/3165839/153068973-99fe4007-9736-40c9-90b0-9c865db0cc33.png">

### Reproduction

https://codesandbox.io/s/react-typescript-forked-pk5z0?file=/src/App.tsx:2269-2334

### Additional context

I noticed this happens when there are nested conditionals and nested states.

**Comment by Valkendorm** [comment]
I found that it happens when the `matches` call is nested in another `matches`. It's also only happening after the first string match, while objects matching will continue to work.

```ts
// This works
if (state.matches('topToo.nested')) return <div>{state.context.stuff}</div>;
if (state.matches('topToo.nestedToo')) return <div>{state.context.stuffToo}</div>;

// This works too
if (state.matches('top')) {
  if (state.matches({ top: 'nested' })) return <div>{state.context.stuff}</div>;
  if (state.matches({ top: 'nestedToo' })) return <div>{state.context.stuffToo}</div>;
}

// "Property 'matches' does not exist on type 'never'." after first string match
if (state.matches('topToo')) {
  if (state.matches('topToo.nested')) return <div>{state.context.stuff}</div>;
  if (state.matches('topToo.nestedToo')) return <div>{state.context.stuffToo}</div>; 
}

// "Property 'matches' does not exist on type 'never'." after first string match
if (state.matches('top')) {
  if (state.matches({ top: 'nestedToo' })) return <div>{state.context.stuffToo}</div>;
  if (state.matches({ top: 'nested' })) return <div>{state.context.stuff}</div>;
  if (state.matches('top.nested')) return <div>{state.context.stuff}</div>;
  if (state.matches('top.nestedToo')) return <div>{state.context.stuffToo}</div>; 
}

```

**Comment by mattpocock** [comment]
@Valkendorm @BrunoQuaresma Are you on latest?

**Comment by Andarist** [comment]
Yes, this still happens with 4.30.0. I've created a minimal repro case here:
[TS playground](https://www.typescriptlang.org/play?noErrorTruncation=true#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs04BfOAMyghDgCIAPAZxizQYG44B6XuDACeYNCwBccACwA6AMwAGGQoBQK5BGJs49ALwp0HPIRJoAFAhVxBLACoixkhJQwtEV64wACX1u2y8wqIA5mjEDJIwUACuaFwe1iBYhGIAyv6OjBgMcAA+WTIARtwe5AA0KuQAlPHqmtpsHHD6IDIkwDDAGAA26Ry1wFRwZo3YMkkwKSxmDNlVcABkC3CjaOPJBGIzGEUMVfOW1qvx1vxwAHoA-JUqQA

The issue is about intersecting `state.matches` return types by using `&&` within the context of the control flow analysis. I will look into this cause conceptually this should just preserve the `this` type (but I recall that we had problems with it since this is a predicate and as such it might use "Exclude" behind the scenes). Perhaps using an overload here would be better because the core problem is that `matches` is currently a type predicate, even when the typegen is used.
